File: src/main/java/com/github/kindrat/cassandra/client/service/tasks/CsvWriteBackgroundTask.java
package com.github.kindrat.cassandra.client.service.tasks;

import com.datastax.driver.core.ColumnMetadata;
import com.datastax.driver.core.TableMetadata;
import com.github.kindrat.cassandra.client.model.CsvTargetMetadata;
import com.github.kindrat.cassandra.client.service.CassandraClientAdapter;
import com.github.kindrat.cassandra.client.service.TableContext;
import com.github.kindrat.cassandra.client.ui.DataObject;
import javafx.collections.ObservableList;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import reactor.core.publisher.Flux;

import java.io.FileWriter;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Function;
import java.util.stream.Collectors;

import static com.github.kindrat.cassandra.client.service.TableContext.fullTable;
import static com.github.kindrat.cassandra.client.service.tasks.TaskState.ACTIVE;

@Slf4j
@RequiredArgsConstructor
public class CsvWriteBackgroundTask extends AbstractTask {
    private final CsvTargetMetadata metadata;
    private final TableMetadata tableMetadata;
    private final CassandraClientAdapter clientAdapter;

    @Override
    Flux<TaskStatistics> innerRun() {
        Instant start = Instant.now();
        return clientAdapter.count(metadata.getTable())
                .flatMapMany(totalRows -> {
                    long percentRows = Math.max(totalRows / 100, 1);
                    percentRows = Math.min(percentRows, Integer.MAX_VALUE);
                    AtomicLong processedRows = new AtomicLong();
                    List<String> columns = tableMetadata.getColumns().stream()
                            .map(ColumnMetadata::getName).collect(Collectors.toList());
                    CSVPrinter printer = getPrinter(columns);
                    TableContext tableContext = fullTable(metadata.getTable(), tableMetadata, clientAdapter, 3000);
                    return Flux.<CompletableFuture<ObservableList<DataObject>>>create(
                            fluxSink -> {
                                fluxSink.next(tableContext.previousPage());
                                while (tableContext.hasNextPage()) {
                                    fluxSink.next(tableContext.nextPage());
                                }
                                fluxSink.complete();
                            })
                            .map(this::extractValue)
                            .flatMapIterable(Function.identity())
                            .doOnNext(dataObject -> printRow(dataObject, columns, printer))
                            .buffer((int) percentRows)
                            .map(dataObjects -> {
                                Duration duration = Duration.between(start, Instant.now());
                                long processed = processedRows.addAndGet(dataObjects.size());
                                double progress = (double) processed / totalRows;
                                return new TaskStatistics(ACTIVE, start.toEpochMilli(), duration, progress);
                            })
                            .doFinally(signalType -> closePrinter(printer));
                })
                .doOnError(throwable -> log.error("Could not write CSV", throwable));
    }

    @SneakyThrows
    private CSVPrinter getPrinter(List<String> columns) {
        CSVFormat format = metadata.getFormat().withHeader(columns.toArray(new String[0]));
        return new CSVPrinter(new FileWriter(metadata.getTarget()), format);
    }

    @SneakyThrows
    private void closePrinter(CSVPrinter printer) {
        printer.close(true);
    }

    @SneakyThrows
    private ObservableList<DataObject> extractValue(CompletableFuture<ObservableList<DataObject>> future) {
        return future.get();
    }

    @SneakyThrows
    private void printRow(DataObject dataObject, List<String> header, CSVPrinter printer) {
        Object[] fields = header.stream().map(dataObject::get).toArray();
        printer.printRecord(fields);
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/service/tasks/CsvWriteBackgroundTask.kt
package com.github.kindrat.cassandra.client.service.tasks

import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata
import com.github.kindrat.cassandra.client.model.CsvTargetMetadata
import com.github.kindrat.cassandra.client.service.CassandraClientAdapter
import com.github.kindrat.cassandra.client.service.TableContext.Companion.fullTable
import com.github.kindrat.cassandra.client.ui.DataObject
import javafx.collections.ObservableList
import mu.KotlinLogging
import org.apache.commons.csv.CSVPrinter
import reactor.core.publisher.Flux
import reactor.core.publisher.FluxSink
import java.io.FileWriter
import java.time.Duration
import java.time.Instant
import java.util.concurrent.CompletionStage
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.max
import kotlin.math.min

private val logger = KotlinLogging.logger {}

class CsvWriteBackgroundTask(
        private val metadata: CsvTargetMetadata,
        private val tableMetadata: TableMetadata,
        private val clientAdapter: CassandraClientAdapter
) : AbstractTask() {

    override fun innerRun(): Flux<TaskStatistics> {
        val start = Instant.now()
        return clientAdapter.count(metadata.table)
                .flatMapMany { totalRows: Long ->
                    var percentRows = max(totalRows / 100, 1)
                    percentRows = min(percentRows, Int.MAX_VALUE.toLong())
                    val processedRows = AtomicLong()
                    val columns = tableMetadata.columns.keys.map { it.toString() }.toList()
                    val printer = getPrinter(columns)
                    val tableContext = fullTable(metadata.table, tableMetadata, clientAdapter, 3000)
                    Flux.create { fluxSink: FluxSink<CompletionStage<ObservableList<DataObject>>> ->
                        fluxSink.next(tableContext.previousPage())
                        while (tableContext.hasNextPage()) {
                            fluxSink.next(tableContext.nextPage())
                        }
                        fluxSink.complete()
                    }
                            .map { extractValue(it) }
                            .flatMapIterable { it }
                            .doOnNext { printRow(it, columns, printer) }
                            .buffer(percentRows.toInt())
                            .map { dataObjects: List<DataObject> ->
                                val duration = Duration.between(start, Instant.now())
                                val processed = processedRows.addAndGet(dataObjects.size.toLong())
                                val progress = processed.toDouble() / totalRows
                                TaskStatistics(TaskState.ACTIVE, start.toEpochMilli(), duration, progress)
                            }
                            .doFinally { closePrinter(printer) }
                }
                .doOnError { logger.error("Could not write CSV", it) }
    }

    private fun getPrinter(columns: List<String>): CSVPrinter {
        val format = metadata.format.withHeader(*columns.toTypedArray())
        return CSVPrinter(FileWriter(metadata.target), format)
    }

    private fun closePrinter(printer: CSVPrinter) {
        printer.close(true)
    }

    private fun extractValue(future: CompletionStage<ObservableList<DataObject>>): ObservableList<DataObject> {
        return future.toCompletableFuture().get()
    }

    private fun printRow(dataObject: DataObject, header: List<String>, printer: CSVPrinter) {
        val fields = header.stream().map { name: String? -> dataObject[name!!] }.toArray()
        printer.printRecord(*fields)
    }
}